### PR TITLE
WIP: Smoltcp socket config

### PR DESCRIPTION
**Pull request by Czocher**
This PR updates the smoltcp version to the newest master one as well as implements the feature specified in #1154 - to allow the dhcpv4 socket to be configured.

Currently it should be considered a WIP PR - requires testing.

@Dirbaio can you have a look and check if this is compatible with what we discussed on the matrix channel?

### CODE REVIEW

File: embassy-net/src/lib.rs [hunk 1]
```diff
@@ -45,25 +47,16 @@ use crate::device::DriverAdapter;
 const LOCAL_PORT_MIN: u16 = 1025;
 const LOCAL_PORT_MAX: u16 = 65535;
 
-pub struct StackResources<const ADDR: usize, const SOCK: usize, const NEIGHBOR: usize> {
-    addresses: [IpCidr; ADDR],
+pub struct StackResources<const SOCK: usize, const NEIGHBOR: usize> {
```
Comment (Dirbaio) [inline embassy-net/src/lib.rs:50]: NEIGHBOR is unsued, could now be removed.

File: embassy-net/src/lib.rs [hunk 2]
```diff
@@ -45,25 +47,16 @@ use crate::device::DriverAdapter;
...
+    addresses: [IpCidr; 5],
```
Comment (Dirbaio) [inline embassy-net/src/lib.rs:51]: `addresses` is now owned by the smoltcp iface, there's no need to have it in this struct (smoltcp iface borrowed it before)

File: embassy-net/src/lib.rs [hunk 3]
```diff
@@ -75,10 +68,34 @@ pub struct Config {
...
+            max_lease_duration: Default::default(),
+            retry_config: Default::default(),
+            ignore_naks: Default::default(),
+            server_port: smoltcp::wire::DHCP_SERVER_PORT,
+            client_port: smoltcp::wire::DHCP_CLIENT_PORT,
+        }
+    }
+}
+
 pub enum ConfigStrategy {
```
Comment (Dirbaio) [inline embassy-net/src/lib.rs:95]: Now that there's config structs for DHCP and static it makes sense to rename like this so they're more consistent:

Config -> StaticConfig
ConfigStrategy -> Config

File: embassy-net/src/lib.rs [hunk 4]
```diff
@@ -105,21 +122,20 @@ impl<D: Driver + 'static> Stack<D> {
     pub fn new<const ADDR: usize, const SOCK: usize, const NEIGH: usize>(
```
Comment (Dirbaio) [inline embassy-net/src/lib.rs:122]: ADDR, NEIGH should be removable.

**Review by Dirbaio — APPROVED** [review]
Thanks for the PR!

You had some auto-formatting tool that had formatted some Cargo.tomls. I've reverted it for you, please make sure it's disabled next time.

I've further updated it to a newer smoltcp and did some more fixes. I've tested it working, so merging.

bors r+

**Comment by Czocher** [comment]
Thanks @Dirbaio sorry, for the autoformat, I'll take that into account next time!

**Pull request merged**